### DISCUSSION TITLE
Excuse me. How do i use this?

### DISCUSSION THREAD

**Discussion by fireisreder** [Elskom/Els_kom_new#178] [Q&A] [answered]
I tried so many times to compile this to no avail.
i gave up then use the legacy version instead..

but here s one thing. i dont know how you re supposed to use this program. i tried copying the .kom files i want to the bin folder. and press unpack. but nothing happen. so can you write a simple tutorial on how to use the program in the read.md many thanks.

**Comment by AraHaan** [comment]
I had one on youtube, basically the program should make a subfolder named ``koms`` and then you copy them there.

Also I have since rebranded the dependencies into a single metapackage, however it has not been pushed to nuget yet, so you would have to clone to the folder to where you cloned this into and then build that using the .NET SDK.

Also I recommend using the .NET 6 SDK from the https://github.com/dotnet/installer/ where on the readme it shows the installer for their "master" branch, this is because the preview1 for .NET 6 branch has a bug in which I reported, even though the one in the .NET 6 branch is newer.

This is because the .NET 5 SDK's that have stable releases has a bug where build would never work as it would not restore the nuget packages so then you would be forced to use the .NET 6 SDK's until VS2019 19.9 is released as stable.

Also there is a bug if you try to compile this using ``dotnet build`` and so would have to build this inside of Visual Studio due to the fact that for some reason the build fails even though it is targeting .NET Framework that it thinks I should reference ``System.Resources.Extensions`` for non-string resources even though those are built into System.Windows.Forms.dll on the .NET Framework.

Also I recommend using the .NET Framework build for now as there is a bug with the positions of the control buttons at the top of the settings form. I been looking into that issue for months.

**Comment by AraHaan** [comment]
So this is the steps in which to be able to build:

run the following commands:
- ``git clone https://github.com/Elskom/Sdk.git --recursive``
- ``cd Sdk``
- ``dotnet build --configuration Release`` (will restore packages needed to build the Sdk).
- ``dotnet pack --configuration Release --no-build``

And make sure the package outputs are added to your pc's global nuget.config and enabled.

Now run:
- `` cd ..`` (to leave the Sdk dir inside of whatever terminal you use (powershell or command prompt)
- ``cd Els_kom_new``
- ``dotnet build --configuration Release`` (will successfully build .NET 5, however it will not work for .NET Framework until you open the solution pcbuild.sln inside of Visual Studio 2019.

Also you will have to pull changes to this repository as when you cloned it did not contain changes I had not pushed yet.

**Comment by AraHaan** [comment]
Also I plan to rename branch master to ``main`` as well and also to remove appveyor.

**Comment by AraHaan** [comment]
Also forgot to mention that the plugins to komv2 and komv3 (that implements packing and unpacking) are currently private at the moment so out of the box they won't be unpack or even packable until they go public again.

The public versions will call in to encryption plugins and I cannot guarantee that the order in algorithm 2 or 3 is correct either. One would have to look at the game source to know for sure.

**Comment by fireisreder** [comment]
thanks for the generous reply. im really interested in the project. all other unpacker like QUICKBMS. Only have version 1 to 4. But in the description this program can unpack ver 5 which is the version of gc im playing. one thing, Is the header editable? like for example the one KOG makes have KOG... header but the private server that i played in has GCM... as the header.

**Comment by AraHaan** [comment]
Well it only supports the kow with KOG on the plugins for now, the GCM ones are the same, they just changed the KOG part to GCM last time I checked.

The komv4 plugin has issues too, for starters it crashes when trying to decrypt the xml header for some odd reason too.

**Comment by AraHaan** [comment]
Also on the about form's code it points to an voidels forms topic on how to use the program too (which had the links to my videos on youtube on the how to use as well too), which has long since been gone 😭 and so the link in the about form been broken since void closure in 2019.

I guess what *should* be done is to open a wiki, however I sadly do not contain the contents of that forum topic or remember exactly what all it said before either which is also sad.

**Comment by AraHaan** [comment]
@shalola I would pull the changes from master, also will rename master too.

**Comment by fireisreder** [comment]
hello, yeah im really new to modding GC. and i did tried the legacy ones. sadly cant decrypt the GCM files.. either bcs of the header or the version of the encryption. gcm uses version 5.

![image](https://user-images.githubusercontent.com/37476685/108004519-60b45100-6ff6-11eb-968b-b6debf95b288.png)

Is it compile-able now? last time i tried it said this error in particular.

**Reply by AraHaan** [reply]
Converted the issue to a discussion.

**Comment by AraHaan** [comment]
It should be able to be compiled, note: I am also working on a copy where the Sdk uses DependencyInjection to fully ensure that it only uses 1 instance of HttpClient for the entire application's lifetime (as if it did not it would be a bug that would exhaust all of the sockets related to the internet on windows which would not be good). The HttpClient however is only used by the plugin update checker which can then install plugin updates (or in some cases uninstall them).

I am also looking into being able to install and uninstall plugins and load/unload them when updated/installed/uninstalled too so that way the program would not need to be restarted as well.

There is also a few minor UI issues like scrollbars not being dark like the ones from Visual Studio Dark theme or the windows dark mode in Windows 10.

As such I am also debating if I should refactor the ListView I use into an control that would do the edit boxes stuff still but then still be able to recolor everything else as well (and basically then remake the forms using WPF).

**Comment by fireisreder** [comment]
![image](https://user-images.githubusercontent.com/37476685/108447374-0ae2e180-7260-11eb-9196-a965daa31025.png)
do you know what happened here? i made sure to add the sdks to the nuget.config file. and this happen.

edit: i tried the other route. and this happen...
![image](https://user-images.githubusercontent.com/37476685/108447860-ef2c0b00-7260-11eb-9b81-1bffd2baa433.png)
![image](https://user-images.githubusercontent.com/37476685/108447895-04089e80-7261-11eb-8bc2-0e17c27956e4.png)
![image](https://user-images.githubusercontent.com/37476685/108447933-11be2400-7261-11eb-8d16-0fe19c30f4e5.png)
![image](https://user-images.githubusercontent.com/37476685/108447956-184c9b80-7261-11eb-8c8f-291ae3a2e4f6.png)

**Reply by AraHaan** [reply]
The hell, it loads without asking for upgrades on my side. Also Els_kom has an issue with building using dotnet build due to a bug in the .NET SDK.

**Reply by fireisreder** [reply]
> The hell, it loads without asking for upgrades on my side. Also Els_kom has an issue with building using dotnet build due to a bug in the .NET SDK.

![image](https://user-images.githubusercontent.com/37476685/108458713-771c1000-7275-11eb-9ce6-6b1f6de942e4.png)

i think i know why. please look at this. if i remember correctly, windows 10 sdk version on your end is lower than this right?

i tried to find your version but in the visual studio installer your version no longer exist so im kinda stuck.. smh for microsoft.

edit: for your reference here s windows 10 sdk currently available
![image](https://user-images.githubusercontent.com/37476685/108459122-5b653980-7276-11eb-94e9-303fa51476bf.png)

**Reply by AraHaan** [reply]
I use the preview visual studio 2019 with the latest 19041 windows sdk.

**Reply by AraHaan** [reply]
Also I had to use an .NET 6 preview SDK for it to work inside of visual studio to build as well from their latest master branch build.

**Reply by fireisreder** [reply]
> I use the preview visual studio 2019 with the latest 19041 windows sdk.

![image](https://user-images.githubusercontent.com/37476685/108511511-4f9e6500-72c0-11eb-9839-9177a36a8d29.png)
![image](https://user-images.githubusercontent.com/37476685/108511543-5af19080-72c0-11eb-9fc0-5dd9289290b5.png)
![image](https://user-images.githubusercontent.com/37476685/108511565-62189e80-72c0-11eb-8f25-a551b5e493dc.png)
![image](https://user-images.githubusercontent.com/37476685/108511599-6c3a9d00-72c0-11eb-82ed-bcb35061b06b.png)
its still happening... I install the dotnet 6 sdk preview too and still got the same problem..

**Reply by AraHaan** [reply]
Some of the issues is that you might not have the latest copy of the repository.

Also I have updated a few things as well with the Sdk which break build since within the last week and have not got around to fix it on this repository yet to update with the breaking changes I done to fix issues within the dependencies that this code depends on.

Due to the Sdk is not fully ready for nuget it can change any any time when issues are found or reported.

Some of the versions inside of the Sdk had lower version numbers, now some of them are in v2 due to breaking changes to fix the issues.

One issue was the plugin loaders for example, before the plugins would get loaded in the default appdomain, which them makes them unable to be reloaded forcing the need to restart the application. Now I made it where on .NET Framework builds it makes an appdomain (or on .NET 5 builds it makes an AssemblyLoadContext) for each plugin and adds them to a list that it has as an output parameter, which then can be used to unload the assemblies without needing to restart the application. Think of it as a plugin container if you want.

Another issue was checking for plugin updates, for starters how it was setup, how do you know if it makes a lot of HttpClient instances instead of only 1 for the lifetime of the process? That is why I had to break it to pass in an ServiceProvider instance which holds said HttpClient instance to ensure such bug does not exist (the application would be responsible for making an ServiceProvider that contains an HttpClient instance) and it would then get easily cleaned up when the application terminates for any reason.

Also unlike before, if plugins have unhandled exceptions, they could be handled and then the plugins safely unloaded without crashing the entire application (similar to what Visual Studio does with extensions).

So ye I also have been working on issues like that as well too.

**Comment by AraHaan** [comment, marked as answer]
The project should build now that I converted the Elskom.Sdk PackageReference to Sdk style.

This should expose how to build properly for now until the next Sdk changes breaks something.

https://github.com/Elskom/Els_kom_new/blob/main/.github/workflows/dotnetcore.yml#L12-L42

Also eventually I plan to do something to push the program to be .NET 5 only in the future and that would mean all plugins would eventually need to target .NET 5 also.

However for now I would use the build on net472 since it is the most consistent at the moment and that would include targeting plugins to net472 currently.

That also means plugins must also be SDK style csproj or SDK style vbproj projects.

And on the projects then must do this on the very first line in the project files:
```xml
<Project Sdk="Elskom.Sdk/1.0.0">
```

When the Sdk is published to nuget.org.
